cargo-shuttle/src/init.rs

```diff
@@ -1,56 +1,291 @@
+    // rules are accepted here. There is no need to check that the user
+    // actually provided a name that the vendor would accept.
+    const GIT_PATTERN: &str = "^(?:(?<vendor>gh|gl|bb):)?(?<owner>[^/.:]+)/(?<name>[^/.:]+)$";
+
+    fn from(loc: TemplateLocation) -> Result<Self> {
+        let git_re = Regex::new(Self::GIT_PATTERN).unwrap();
+
+        if let Some(caps) = git_re.captures(&loc.auto_path) {
+            let vendor = match caps.name("vendor").map(|v| v.as_str()) {
+                Some("gl") => GitVendor::GitLab,
+                Some("bb") => GitVendor::BitBucket,
+                // GitHub is the default vendor if no other vendor is specified.
+                Some("gh") | None => GitVendor::GitHub,
+                Some(_) => unreachable!("should never match unknown vendor"),
+            };
+
+            Ok(Self::RemoteRepo {
+                vendor,
+                // `owner` and `name` are required for the regex to
+                // match. Thus, we don't need to check if they exist.
+                owner: caps["owner"].to_owned(),
+                name: caps["name"].to_owned(),
+                subfolder: loc.subfolder.map(PathBuf::from),
+            })
+        } else if Path::new(&loc.auto_path).is_absolute() || loc.auto_path.starts_with('.') {
+            // Local paths to template locations must be absolute or start with a
+            // dot pattern. This way, conflicts between relative path in the form
+            // `foo/bar` and repository name in the form `owner/name` are avoided.
+            match loc.subfolder {
+                Some(subfolder) => Ok(Self::LocalPath(Path::new(&loc.auto_path).join(subfolder))),
+                None => Ok(Self::LocalPath(PathBuf::from(loc.auto_path))),
+            }
+        } else if let Ok(url) = loc.auto_path.parse::<Url>() {
+            if url.scheme() == "http" || url.scheme() == "https" {
+                Ok(Self::Url {
+                    url: loc.auto_path,
+                    subfolder: loc.subfolder.map(PathBuf::from),
+                })
+            } else {
+                println!(
+                    "URL scheme is not supported. Please use HTTP of HTTPS for URLs\
+		     , or use another method of specifying the template location."
+                );
+                println!(
+                    "HINT: Here you can find examples of how to \
+		     select a template: https://github.com/shuttle\
+		     -hq/shuttle-examples#how-to-clone-run-and-deploy-an-example"
+                );
+                anyhow::bail!("invalid URL scheme")
+            }
+        } else {
+            anyhow::bail!("template location is invalid")
+        }
+    }
+
+    fn generate<F>(self, dest: &Path, prepare: F) -> Result<()>
+    where
+        F: FnOnce(&Path) -> Result<()>,
+    {
+        let temp_dir =
+            tempdir().context("Failed to create a temporary directory to generate project into")?;
```

Comment:
Can you use a prefix, like we do in the integration tests?

Reply:
What prefix would you suggest? I'm torn between `shuttle-init`, which is shorter, and `cargo-shuttle-init`, which makes the context more clear.